com.unity.render-pipelines.core/Editor/SpeedTree8MaterialUpgrader.cs

```diff
@@ -0,0 +1,181 @@
+            "_WINDQUALITY_PALM"
+        };
+
+        /// <summary>
+        /// Creates a material upgrader with only the renames in common between HD and Universal.
+        /// </summary>
+        /// <param name="sourceShaderName">Original ST8 shader name.</param>
+        /// <param name="destShaderName">New ST8 shader name.</param>
+        /// <param name="finalizer">A delegate that postprocesses the material as needed by the render pipeline.</param>
+        public SpeedTree8MaterialUpgrader(string sourceShaderName, string destShaderName, MaterialFinalizer finalizer = null)
+        {
+            RenameShader(sourceShaderName, destShaderName, finalizer);
+            RenameFloat("_WindQuality", "_WINDQUALITY");
+            RenameFloat("_TwoSided", "_CullMode"); // Currently only used in HD. Update this once URP per-material cullmode is enabled via shadergraph. 
+        }
+
+        private static void ImportNewSpeedTree8Material(Material mat, int windQuality, bool isBillboard)
+        {
+            int cullmode = 0;
+            mat.SetFloat("_WINDQUALITY", windQuality);
+            if (isBillboard)
+            {
+                mat.EnableKeyword("EFFECT_BILLBOARD");
+                cullmode = 2;
+            }
+            if (mat.HasProperty("_CullMode"))
+                mat.SetFloat("_CullMode", cullmode);
+        }
+
+        /// <summary>
+        /// Postprocess materials when importing a SpeedTree8 asset. Call from OnPostprocessSpeedTree in a MaterialPostprocessor.
+        /// </summary>
+        /// <param name="speedtree">Game object for the SpeedTree asset being imported.</param>
+        /// <param name="stImporter">The assetimporter used to import the SpeedTree asset.</param>
+        /// <param name="finalizer">Render pipeline-specific material finalizer.</param>
+        public static void PostprocessSpeedTree8Materials(GameObject speedtree, SpeedTreeImporter stImporter, MaterialFinalizer finalizer = null)
+        {
+            LODGroup lg = speedtree.GetComponent<LODGroup>();
+            LOD[] lods = lg.GetLODs();
+            for (int l = 0; l < lods.Length; l++)
+            {
+                LOD lod = lods[l];
+                bool isBillboard = stImporter.hasBillboard && (l == lods.Length - 1);
+                int wq = Mathf.Min(stImporter.windQualities[l], stImporter.bestWindQuality);
+                foreach (Renderer r in lod.renderers)
+                {
+                    foreach (Material m in r.sharedMaterials)
+                    {
+                        ImportNewSpeedTree8Material(m, wq, isBillboard);
+                        if (finalizer != null)
+                            finalizer(m);
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Preserve wind quality and billboard setting when upgrading a ST8 material from previous versions.
+        /// Wind priority order is enabled keyword > _WindQuality float value.
```

Comment:
What is "wind priority"? 
I think the sentence would go something like: "To enable wind priority, use the _WindQuality keyword . . . "
but it isn't clear how the float value relates to this.

Reply:
This code looks like it could be related, defining the _WindQuality keyword . . . 

`windQuality = material.HasProperty("_WindQuality") ? (int)material.GetFloat("_WindQuality") : 0;`

Reply:
Wind priority here refers to what the code prioritizes as the "authority" when guessing what WindQuality to set to a material. 

First it checks the keywords- if a quality is set there, it's the source of truth. Otherwise, we check the value set to the float property. 

This is only relevant because HD does some keyword resetting when synchronizing its materials.

Reply:
Thanks!
So how does the float value determine the WindQuality? If the float is higher presumably it has a specific effect, and if it is lower presumably it has a different effect.

Progressing in the rewrite, but don't yet have all the info . . . 
To determine which WindQuality to use, Unity checks the keywords first and then the _WindQuality float value. [explain about float value here].

Reply:
Wind Quality is defined by speed tree. Speed Tree has documentation here: 
https://docs8.speedtree.com/modeler/doku.php?id=windunity

Reply:
Oh right, not Unity-specific. That said, the SpeedTree doc doesn't clarify which float values correspond with which quality levels and I can't seem to find that info anywhere. It is important that folks know what the float value means.

Reply:
So we should probably add it to the main SpeedTree documentation page in Unity: 
https://docs.unity3d.com/Manual/SpeedTree.html

The quality settings are:
0 = None
1 = Fastest
2 = Faster
3 = Better
4 = Best
5 = Palm

Reply:
Does the comments for this function need any change though?

Reply:
Ok, I'd propose that I arrange to update the main SpeedTree doc page to add then quality settings.m
With that in mind:

To determine which WindQuality to use, Unity checks the keywords first and then the _WindQuality float value. See SpeedTree in the Unity Manual for the values associated with different WindQuality settings.

Reply:
Added a note to the ticket about updating manual page for speedtree as well, I will do it if not owned by Sharlene.